### ISSUE TITLE
cmd_exec hangs for secure mount when a large amount of mounts are present

### ISSUE THREAD

**Issue by lukehinds** [keylime/keylime#330]
Run the following script:

```
#!/bin/bash
for i in {1..300}
do
   mkdir /tmp/tmpfs_test${i}
   mount -t tmpfs -o size=1m  tmpfs_test${i} /tmp/tmpfs_test${i}
done
```

This should give you 300+ tmpfs mounts (being tmpfs is not so important, its more about having a high amount of stdout for the mount command). 

```
mount | wc -l
336
```

Should you now run keylime agent it will hang when trying to view the current mount command output:

```
keylime_agent
Using config file /etc/keylime.conf
2020-07-15 10:28:00.268 - keylime.tpm2 - INFO - TPM2-TOOLS Version: 3.2.0
^CTraceback (most recent call last):
  File "/usr/local/bin/keylime_agent", line 6, in <module>
    from keylime.cmd.agent import main
  File "/usr/local/lib/python3.7/site-packages/keylime/cmd/agent.py", line 9, in <module>
    from keylime import keylime_agent
  File "/usr/local/lib/python3.7/site-packages/keylime/keylime_agent.py", line 48, in <module>
    tpm = tpm_obj.getTPM(need_hw_tpm=True)
  File "/usr/local/lib/python3.7/site-packages/keylime/tpm/tpm_obj.py", line 83, in getTPM
    __tpm2 = tpm2.tpm2(need_hw_tpm)
  File "/usr/local/lib/python3.7/site-packages/keylime/tpm/tpm2.py", line 78, in __init__
    self.__startup_tpm()
  File "/usr/local/lib/python3.7/site-packages/keylime/tpm/tpm2.py", line 366, in __startup_tpm
    retDict = self.__run("tpm2_startup -c")
  File "/usr/local/lib/python3.7/site-packages/keylime/tpm/tpm2.py", line 266, in __run
    retDict = cmd_exec.run(cmd=cmd, expectedcode=expectedcode, raiseOnError=False, lock=lock, outputpaths=outputpaths, env=env)
  File "/usr/local/lib/python3.7/site-packages/keylime/cmd_exec.py", line 24, in run
    code = proc.wait()
  File "/usr/lib64/python3.7/subprocess.py", line 1019, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib64/python3.7/subprocess.py", line 1653, in _wait
    (pid, sts) = self._try_wait(0)
  File "/usr/lib64/python3.7/subprocess.py", line 1611, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
```

To cleanup:

```
#!/bin/bash
for i in {1..300}
do
   umount tmpfs_test${i}
   rm -rf /tmp/tmpfs_test${i}
done
```

I think we need to approach how we use subprocess, perhaps update to using subprocess.communicate instead of wait.

**Comment by lukehinds** [comment]
@AyushAmbastha @mpeters - this is the issue stopping the agent from running on a kubenetes node. 

@jetwhiz any ideas on approaches here?

**Comment by lukehinds** [comment]
This is definitely the issue: https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait

**Comment by nabilschear** [comment]
an alternative idea, can we just leverage an existing tmpfs partition like /run and symlink /var/lib/keylime/secure to somewhere in run?

**Comment by mpeters** [comment]
> an alternative idea, can we just leverage an existing tmpfs partition like /run and symlink /var/lib/keylime/secure to somewhere in run?

That would still leave it vulnerable to the same bug in other contexts: where the output of some command is larger than the internal buffer Popen is using to read it. It's weird that this `mount` command is what triggered it, but it could happen in lots of other contexts on complex systems.

**Comment by mpeters** [comment]
Should be fixed now in v5.7.0
